text/3491-remove-implicit-features.md

```diff
@@ -0,0 +1,129 @@
+- Feature Name: `remove-implicit-feature`
+- Start Date: 2023-09-18
+- RFC PR: [rust-lang/rfcs#3491](https://github.com/rust-lang/rfcs/pull/3491)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+By default, cargo will treat any optional dependency as a
+[feature](https://doc.rust-lang.org/cargo/reference/features.html).
+As of cargo 1.60, these can be disabled by declaring a feature that activates
+the optional dependency as `dep:<name>`
+(see [RFC #3143](https://rust-lang.github.io/rfcs/3143-cargo-weak-namespaced-features.html)).
+
+On the next edition, cargo will stop exposing optional dependencies as features
+implicitly, requiring users to add `foo = ["dep:foo"]` if they still want it exposed.
+
+# Motivation
+[motivation]: #motivation
+
+While implicit features offer a low overhead way of defining features,
+- It is easy to overlook using `dep:` when the optional dependency is not intended to be exposed
+  - Making it easy for crate authors to use the wrong syntax and be met with errors ([rust-lang/cargo#10125](https://github.com/rust-lang/cargo/issues/10125))
+  - Potentially breaking people if they are later removed ([rust-lang/cargo#12687)](https://github.com/rust-lang/cargo/pull/12687))
+  - Leading to confusing choices when `cargo add` lists features that look the same (e.g. `cargo add serde` showing `derive` and `serde_derive`)
+  - Leading to confusing errors for callers when they reference the dependency, instead of the feature, and things don't work right
+- Tying feature names to dependency names is a code smell because it ties the API to the implementation
+- It requires people and tools to deal with this special case ([rust-lang/cargo#10543](https://github.com/rust-lang/cargo/issues/10543))
+  - Granted, anything having to deal with old editions will still have to deal with this
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+
+Updated documentation:
+
+## Optional Dependencies
+*(Replaces the [same section in the book](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies))*
+
+Dependencies can be marked "optional", which means they will not be compiled
+by default.
+
+For example, let's say that our 2D image processing library uses
+an external package to handle GIF images. This can be expressed like this:
+
+```toml
+[features]
+gif = ["dep:giffy"]
+
+[dependencies]
+giffy = { version = "0.11.1", optional = true }
+```
```

Comment:
I assume `gif = ["dep:gif"]` (matching names) also works? if so, maybe this should be explicitly mentioned or even used as the example to avoid people from thinking that this would not be allowed?

Reply:
Yes, it does.

I wrote it from the perspective of someone reading it in 5 years where there is no historical baggage around assuming dependencies and features are in the same namespace.